### PR TITLE
Stream report downloads incrementally

### PR DESCRIPTION
**Pull request by SamWebmecanik**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | ✔️ 
| New feature? | 
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Exporting reports used to run out of memory quickly. This PR aims to stream the report downloads incrementally so as to be able to export arbitrarily large reports without crashing Mautic.

So far, this has only been done for the CSV exports and a lot of the logic is in the controller.
I would greatly appreciate any feedback and help completing this fix.

BTW, I have no clue how to do the same for the excel export.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1.  Make a big report
2. Try to export it
3. Hopefully make Mautic run out of memory (if not, either try to make your report bigger or lower the PHP memory limit)

#### Steps to test this PR:
1.  Open that same big report
2. Try to export it as a CSV
3. The memory should not run out

### CODE REVIEW

File: app/bundles/ReportBundle/Controller/ReportController.php [hunk 1]
```diff
@@ -783,14 +783,43 @@ public function exportAction($objectId, $format = 'csv')
...
+        $name = str_replace(' ', '_', $date) . '_' . InputHelper::alphanum($entity->getName(), false, '-');
+        $options = array( 'dateFrom' => new \DateTime($fromDate),'dateTo' => new \DateTime($toDate));
 
-        $reportData = $model->getReportData($entity, null, [
-            'dateFrom'       => new \DateTime($fromDate),
-            'dateTo'         => new \DateTime($toDate),
-            'dynamicFilters' => $dynamicFilters,
-        ]);
+        if($format === 'html') {
+            if ($format === 'xlsx') {
```
Comment (escopecz) [inline app/bundles/ReportBundle/Controller/ReportController.php]: These two conditions doesn't make much sense. If $format === 'html', it can never === 'xlsx'

Comment (SamWebmecanik) [inline app/bundles/ReportBundle/Controller/ReportController.php]: yes, I change it.

**Comment by SamWebmecanik** [comment]
I find a possible optimization to display report page in SQL query but I don't know where I should make the modification.

Origin request
`SELECT e.id, e.subject as title, count(CASE WHEN es.is_read THEN 1 ELSE null END) as "read" FROM
email_stats es LEFT JOIN emails e ON e.id = es.email_id LEFT JOIN leads l ON l.id = es.lead_id LEFT JOIN
ip_addresses i ON i.id = es.ip_id LEFT JOIN emails vp ON vp.id = e.variant_parent_id LEFT JOIN categories
c ON c.id = e.category_id WHERE (es.date_sent BETWEEN '2017-02-05 00:00:00' AND '2017-03-16
15:48:23') AND (e.is_published = 1) GROUP BY e.id, e.subject  ORDER BY "read" DESC LIMIT 10 OFFSET 0`

New request
`SELECT e.id, e.subject as title, count(CASE WHEN es.is_read THEN 1 ELSE null END) as "read" FROM email_stats es LEFT JOIN emails e ON e.id = es.email_id LEFT JOIN leads l ON l.id = es.lead_id LEFT JOIN ip_addresses i ON i.id = es.ip_id LEFT JOIN emails vp ON vp.id = e.variant_parent_id LEFT JOIN categories c ON c.id = e.category_id WHERE (es.date_sent BETWEEN '2017-02-05 00:00:00' AND '2017-03-16 15:48:23') AND (e.is_published = 1) GROUP BY e.id ORDER BY "read" DESC LIMIT 10 OFFSET 0`

To optimize this query, we may delete second parameter of the 'Group by' : e.subject.
This parameter is useless because we already use `e.id` for group by consequently `e.subject` doesn't make a better job.

**Comment by npracht** [comment]
Hi @escopecz would you consider adding that PR in 2.8.0 ?
I saw that new PR has been attached to 2.8.0 but ours (which is older) not :(

Tell us how to help for that.

N.

File: app/bundles/ReportBundle/Controller/ReportController.php [hunk 2]
```diff
@@ -783,14 +783,42 @@ public function exportAction($objectId, $format = 'csv')
...
-            'dateTo'         => new \DateTime($toDate),
-            'dynamicFilters' => $dynamicFilters,
-        ]);
+            $response->headers->set('Content-Type', 'application/force-download');
+            $response->headers->set('Content-Type', 'application/octet-stream');
+            $response->headers->set('Content-Disposition', 'attachment; filename="' . $name . '.'.$format.'"');
+            $response->headers->set('Expires', 0);
+            $response->headers->set('Cache-Control', 'must-revalidate');
+            $response->headers->set('Pragma', 'public');
+        }else{
```
Comment (escopecz) [inline app/bundles/ReportBundle/Controller/ReportController.php]: Please check the Mautic coding standard and the pre-commit hook:

https://github.com/mautic/mautic/blob/staging/.github/CONTRIBUTING.md#coding-standards

There are also other places where your code is not following the standard.

File: app/bundles/ReportBundle/Model/ReportModel.php [hunk 3]
```diff
@@ -447,37 +423,31 @@ function () use ($formatter, $reportData, $report, $name) {
...
                         }
                     );
-
                     $response->headers->set('Content-Type', 'application/force-download');
                     $response->headers->set('Content-Type', 'application/octet-stream');
                     $response->headers->set('Content-Disposition', 'attachment; filename="'.$name.'.xlsx"');
                     $response->headers->set('Expires', 0);
                     $response->headers->set('Cache-Control', 'must-revalidate');
                     $response->headers->set('Pragma', 'public');
-
```
Comment (escopecz) [inline app/bundles/ReportBundle/Model/ReportModel.php]: Is there need to remove these lines?

File: app/bundles/ReportBundle/Model/ReportModel.php [hunk 4]
```diff
@@ -363,80 +363,56 @@ public function getGraphList($context = 'all')
...
+                        'columns' => $reportData['columns'],
                         'pageTitle' => $name,
-                        'graphs'    => $reportData['graphs'],
-                        'report'    => $report,
-                        'dateFrom'  => $reportData['dateFrom'],
-                        'dateTo'    => $reportData['dateTo'],
+                        'graphs' => $reportData['graphs'],
+                        'report' => $report,
+                        'dateFrom' => $reportData['dateFrom'],
+                        'dateTo' => $reportData['dateTo'],
```
Comment (escopecz) [inline app/bundles/ReportBundle/Model/ReportModel.php]: Please return the indentation back here and on other places. Also please revert the changes from `'.'` to `' . '`

**Review by escopecz — CHANGES_REQUESTED** [review]
Hi, thanks for this PR! I noticed few small issues at the first glance.

File: app/bundles/ReportBundle/Controller/ReportController.php (hunk 2 above)
Comment (SamWebmecanik) [inline app/bundles/ReportBundle/Controller/ReportController.php]: Done

Comment (escopecz) [inline app/bundles/ReportBundle/Controller/ReportController.php]: Thanks!, but I don't see it all fixed. `}else{` should be `} else {`. `if(` should be `if (`. `array( '` should be `array('`

File: app/bundles/ReportBundle/Model/ReportModel.php [hunk 5]
```diff
@@ -363,55 +363,32 @@ public function getGraphList($context = 'all')
      *
      * @throws \Exception
      */
-    public function exportResults($format, $report, $reportData)
+public function exportResults($format, $report, $reportData, $handle, $page)
```
Comment (escopecz) [inline app/bundles/ReportBundle/Model/ReportModel.php]: This line should be indented by 4 spaces to the right. A problem is that by adding these additional mandatory parameters you broke BC. The method can be used in 3rd party plugins and this will break them.

**Review by escopecz — CHANGES_REQUESTED** [review]
One more round :)

File: app/bundles/ReportBundle/Model/ReportModel.php (hunk 5 above)
Comment (SamWebmecanik) [inline app/bundles/ReportBundle/Model/ReportModel.php]: Ok I will change it

**Comment by SamWebmecanik** [comment]
Done

**Comment by escopecz** [comment]
Could you check occurrences of `if(` one more time, please?

File: app/bundles/ReportBundle/Model/ReportModel.php [hunk 6]
```diff
@@ -716,4 +701,24 @@ private function getOperatorOptions(array $data)
 
         return $options;
     }
+
+    private function exportCSV($formatter,$report, $reportData, $handle, $page){
```
Comment (escopecz) [inline app/bundles/ReportBundle/Model/ReportModel.php]: This line should have a couple more spaces `private function exportCSV($formatter, $report, $reportData, $handle, $page) {`

File: app/bundles/ReportBundle/Model/ReportModel.php [hunk 7]
```diff
@@ -363,55 +363,35 @@ public function getGraphList($context = 'all')
...
-                                fputcsv($handle, $header);
-                            }
-
-                            //write the row
-                            fputcsv($handle, $row);
-
-                            //free memory
-                            unset($row, $reportData['data'][$count]);
+                //build the data rows
+                if(is_null($handle)){
```
Comment (escopecz) [inline app/bundles/ReportBundle/Model/ReportModel.php]: And also a space before `{` here.

File: app/bundles/ReportBundle/Controller/ReportController.php [hunk 8]
```diff
@@ -783,14 +785,42 @@ public function exportAction($objectId, $format = 'csv')
         $fromDate = $session->get('mautic.report.date.from', (new \DateTime('-30 days'))->format('Y-m-d'));
         $toDate   = $session->get('mautic.report.date.to', (new \DateTime())->format('Y-m-d'));
 
-        $dynamicFilters = $session->get('mautic.report.'.$objectId.'.filters', []);
+        $date = (new DateTimeHelper())->toLocalString();
+        $name = str_replace(' ', '_', $date) . '_' . InputHelper::alphanum($entity->getName(), false, '-');
```
Comment (escopecz) [inline app/bundles/ReportBundle/Controller/ReportController.php]: There should not be spaces between `'` and `.`

**Review by escopecz — CHANGES_REQUESTED** [review]
I feel like a CS nazi. Sorry about that. We fight to keep Mautic's code as perfect as it can be.

File: app/bundles/ReportBundle/Model/ReportModel.php (hunk 7 above)
Comment (SamWebmecanik) [inline app/bundles/ReportBundle/Model/ReportModel.php]: Sorry it's my fault and I find why =/

**Comment by SamWebmecanik** [comment]
I think (hope) is all right ^^

**Comment by escopecz** [comment]
It's exporting the report CSV fine for me. The only problem I can see is that it's exporting only data, not the column names as it did before. Could you add it back?

**Comment by SamWebmecanik** [comment]
I will do it today.

**Comment by SamWebmecanik** [comment]
Done ^.^

**Review by escopecz — APPROVED** [review]
Works for me now. Thanks! 👍

**Comment by mqueme** [comment]
Works for me as well, thanks
